[BE] Flaky Unit Tests

### Describe the bug

Every now and then, some backend tests fail. This is an issue to track all of them in an effort to fix the flakyness and make them consistent

### Steps to reproduce

1. Run `mage test:ci` until one of them fails. The failure rate is really low (less than 1 out of 20 times), but it exists

### Expected behavior

Tests should always pass or always fail

### Environment

CircleCI

### Additional context

```
--- FAIL: TestSendDataIfTimeLimitHasBeenReached (1.64s)
    s3_test.go:305: FAIL:	Upload(string,string)
        		at: [s3_test.go:300]
    s3_test.go:305: FAIL: 0 out of 1 expectation(s) were met.
        	The code you are testing needs to make 1 more call(s).
        	at: [s3_test.go:305]
    s3_test.go:306: FAIL:	Publish(string)
        		at: [s3_test.go:301]
    s3_test.go:306: FAIL: 0 out of 1 expectation(s) were met.
        	The code you are testing needs to make 1 more call(s).
        	at: [s3_test.go:306]
```

### Screenshots

<img width="1236" alt="Screen Shot 2020-12-03 at 10 32 55 AM" src="https://user-images.githubusercontent.com/10436045/100984273-2ee13180-3553-11eb-899f-e2ecee6ea79f.png">

--- a/internal/log_analysis/log_processor/destinations/s3.go
+++ b/internal/log_analysis/log_processor/destinations/s3.go
@@ -152,8 +152,6 @@ func (d *S3Destination) SendEvents(parsedEventChannel chan *parsers.Result, errC
 	// accumulate results gzip'd in a buffer
 	failed := false // set to true on error and loop will drain channel
 	bufferSet := d.newS3EventBufferSet()
-	eventsProcessed := 0
-
 	zap.L().Debug("starting to read events from channel")
 	run := true
 	for run {
@@ -212,7 +210,7 @@ func (d *S3Destination) SendEvents(parsedEventChannel chan *parsers.Result, errC
 	close(sendChan)
 	sendWaitGroup.Wait() // wait until all writes to s3 are done
 
-	zap.L().Debug("finished sending s3 files", zap.Int("events", eventsProcessed))
+	zap.L().Debug("finished sending s3 files")
 }
 
 // sendData puts data in S3 and sends notification to SNS

--- a/internal/log_analysis/log_processor/destinations/s3.go
+++ b/internal/log_analysis/log_processor/destinations/s3.go
@@ -153,13 +153,10 @@ func (d *S3Destination) SendEvents(parsedEventChannel chan *parsers.Result, errC
 	failed := false // set to true on error and loop will drain channel
 	bufferSet := d.newS3EventBufferSet()
 	eventsProcessed := 0
-	zap.L().Debug("starting to read events from channel")
-	for event := range parsedEventChannel {
-		if failed { // drain channel
-			continue
-		}
 
-		// Check if any buffer has held data for longer than maxDuration
+	zap.L().Debug("starting to read events from channel")
+	run := true
+	for run {
 		select {
 		case <-flushExpired.C:
 			now := time.Now() // NOTE: not the same as the tick time which can be older
@@ -170,34 +167,43 @@ func (d *S3Destination) SendEvents(parsedEventChannel chan *parsers.Result, errC
 				}
 				sendChan <- tooOldBuffer
 			}
-		default: // makes select non-blocking
-		}
-		sendBuffers, err := bufferSet.writeEvent(event)
-		if err != nil {
-			failed = true
-			zap.L().Debug(`aborting log processing: failed to write event`, zap.Error(err), zap.String(`logType`, event.PantherLogType))
-			errChan <- errors.Wrapf(err, "failed to write event %s", event.PantherLogType)
-			continue
-		}
-		// buffers needs flushing
-		for _, buf := range sendBuffers {
-			sendChan <- buf
-		}
+		case event, ok := <-parsedEventChannel:
+			if !ok {
+				// If channel has been closed,
+				// stop the loop
+				run = false
+				break
+			}
 
-		eventsProcessed++
+			if failed {
+				// If we have encountered already a failure
+				// just ignore the messages
+				break
+			}
+
+			sendBuffers, err := bufferSet.writeEvent(event)
+			if err != nil {
+				failed = true
+				zap.L().Debug(`aborting log processing: failed to write event`, zap.Error(err), zap.String(`logType`, event.PantherLogType))
+				errChan <- errors.Wrapf(err, "failed to write event %s", event.PantherLogType)
+				continue
+			}
+			// buffers needs flushing
+			for _, buf := range sendBuffers {
+				sendChan <- buf
+			}
+		}
 	}
 
-	if failed {
-		zap.L().Debug("failed, returning after draining parsedEventsChannel")
+	if !failed {
+		zap.L().Debug("output channel closed, sending last events")
+		// If the channel has been closed send the buffered messages before terminating
+		_ = bufferSet.apply(func(buffer *s3EventBuffer) (bool, error) {
+			bufferSet.removeBuffer(buffer) // bufferSet is not thread safe, do this here
+			sendChan <- buffer
+			return false, nil
+		})
 	}
-
-	zap.L().Debug("output channel closed, sending last events")
-	// If the channel has been closed send the buffered messages before terminating
-	_ = bufferSet.apply(func(buffer *s3EventBuffer) (bool, error) {
-		bufferSet.removeBuffer(buffer) // bufferSet is not thread safe, do this here
-		sendChan <- buffer
-		return false, nil
-	})
 
 	// FIXME: closing the channel here is appropriate but we risk a panic leaving the write goroutine open forever.
 	// causing a memory leak. To fix this we need to have the reading of results in a goroutine and keep the writing here.

--- a/internal/log_analysis/log_processor/destinations/s3.go
+++ b/internal/log_analysis/log_processor/destinations/s3.go
@@ -174,8 +174,9 @@ func (d *S3Destination) SendEvents(parsedEventChannel chan *parsers.Result, errC
 			}
 
 			if failed {
-				// If we have encountered already a failure
-				// just ignore the messages
+				// If we have encountered already a failure just ignore the messages
+				// Note that we don't stop when we encounter a failure but use this control variable
+				// since we want to make sure we drain the `parsedEventChannel`
 				break
 			}
 

--- a/internal/log_analysis/log_processor/destinations/s3_test.go
+++ b/internal/log_analysis/log_processor/destinations/s3_test.go
@@ -23,6 +23,7 @@ import (
 	"compress/gzip"
 	"io/ioutil"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -282,23 +283,29 @@ func TestSendDataIfTimeLimitHasBeenReached(t *testing.T) {
 	t.Parallel()
 
 	destination := mockDestination()
-	destination.maxDuration = 200 * time.Millisecond
+	destination.maxDuration = 50 * time.Millisecond
 
 	const nevents = 4
+	var wg sync.WaitGroup
+
 	eventChannel := make(chan *parsers.Result, nevents)
 	go func() {
 		defer close(eventChannel)
 		for i := 0; i < nevents; i++ {
+			wg.Add(1)
 			eventChannel <- newSimpleTestEvent().Result()
 			// The destination writes events to S3 every 'maxDuration' time.
-			// While we sleep here, the destination should write to S3
-			// the event we just wrote to the eventChannel
-			time.Sleep(2 * destination.maxDuration)
+			// Wait until it has published the message to SNS to send the next message
+			wg.Wait()
 		}
 	}()
 
 	destination.mockS3Uploader.On("Upload", mock.Anything, mock.Anything).Return(&s3manager.UploadOutput{}, nil).Times(nevents)
-	destination.mockSns.On("Publish", mock.Anything).Return(&sns.PublishOutput{}, nil).Times(nevents)
+	destination.mockSns.On("Publish", mock.Anything).Return(&sns.PublishOutput{}, nil).Run(func(args mock.Arguments) {
+		// Every time we publish a message to SNS,
+		// unblock the loop that sends messages to the eventsChannel
+		wg.Done()
+	}).Times(nevents)
 
 	assert.NoError(t, runDestination(destination, eventChannel))
 

